leveldb/session_util.go

```diff
@@ -39,20 +39,155 @@ func (s *session) newTemp() storage.FileDesc {
 	return storage.FileDesc{Type: storage.TypeTemp, Num: num}
 }
 
-func (s *session) addFileRef(fd storage.FileDesc, ref int) int {
-	ref += s.fileRef[fd.Num]
+// addFileRef adds file reference counter with specified file number and
+// reference value; need external synchronization.
+func (s *session) addFileRef(fnum int64, ref int) int {
+	ref += s.fileRef[fnum]
 	if ref > 0 {
-		s.fileRef[fd.Num] = ref
+		s.fileRef[fnum] = ref
 	} else if ref == 0 {
-		delete(s.fileRef, fd.Num)
+		delete(s.fileRef, fnum)
 	} else {
-		panic(fmt.Sprintf("negative ref: %v", fd))
+		panic(fmt.Sprintf("negative ref: %v", fnum))
 	}
 	return ref
 }
 
 // Session state.
 
+const cachedTaskBound = 256
+
+// vDelta indicates the change information between the next version
+// and the currently specified version
+type vDelta struct {
+	vid     int64
+	added   []int64
+	deleted []int64
+}
+
+// vTask defines a version task for either reference or release.
+type vTask struct {
+	vid   int64
+	files []tFiles
+}
+
+func (s *session) refLoop() {
+	var (
+		ref        = make(map[int64][]tFiles)
+		deltas     = make(map[int64]*vDelta)
+		referenced = make(map[int64]struct{})
+		released   = make(map[int64]*vDelta)
+		next, last int64
+	)
+	// processTasks processes version tasks in strict order.
+	//
+	// If we want to use delta to reduce the cost of file references and dereferences,
+	// we must strictly follow the id of the version, otherwise some files that are
+	// being referenced will be deleted.
+	//
+	// In addition, some db operations (such as iterators) may cause a version to be
+	// referenced for a long time. In order to prevent such operations from blocking
+	// the entire processing queue, we will properly convert some of the version tasks
+	// into full file references and releases.
+	processTasks := func() {
+		// Make sure we don't cache too many version tasks.
+		for {
+			if last-next < cachedTaskBound {
+				break
+			}
+			if _, exist := released[next]; exist {
+				break
+			}
+			if _, exist := ref[next]; !exist {
+				break
+			}
+			for _, tt := range ref[next] {
+				for _, t := range tt {
+					s.addFileRef(t.fd.Num, 1)
+				}
+			}
+			referenced[next] = struct{}{}
+			delete(ref, next)
+			delete(deltas, next)
+			next += 1
+		}
+
+		// Use delta information to process all released versions.
+		for {
+			if d, exist := released[next]; exist {
+				if d != nil {
```

Comment:
can `d` be `nil` if `exist` is true?

Reply:
Yes, for `read` operation, it just references the corresponding version and then releases when read is done. No `delta` introduced.